Make RandomWindowGeoDataset more efficient when sampling chips from scenes with AOIs

AdeelH opened the pull request:
## Overview

Makes the sampling of random windows from scenes with AOIs more efficient. The previous approach was to sample a random window from anywhere within the scene and check to see if it falls within the AOI. In contrast, this approach is able to generate random window locations that are always inside the AOI. We still need to check if the window is fully contained within the AOI, since sampled windows near the boundaries of the AOI will occasionally overflow the AOI.

The algorithm can be summarized as:
- Partition the AOI polygons into triangles (triangulation).
    - This procedure is also able to account for holes inside polygons (see example below).
- When sampling, pick a triangle (`ABC`) at random with probability proportional to its area.
- Starting at `A`, travel a random distance along vectors `AB` and `AC`.
- Return the final position.

### Visualizations

#### Example AOI
![aoi](https://user-images.githubusercontent.com/13014700/127647160-76ffc1ad-ab80-4256-9959-cc042daeffea.png)

#### Old sampling approach
![random_chips](https://user-images.githubusercontent.com/13014700/127647177-1ccf6f85-a354-4723-a5fb-cd599a03f47d.png)

#### New sampling approach
##### triangulation
![triangulation](https://user-images.githubusercontent.com/13014700/127646818-123b06e3-81eb-4f82-9ee8-600d0dc4ca20.png)
##### sampled windows
![aoi_chips](https://user-images.githubusercontent.com/13014700/127647191-7a700a7b-9002-4fc9-8e29-ae1845177add.png)


### Checklist

- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [x] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

### Notes

- New dependency added: [triangle](https://github.com/drufat/triangle)

## Testing Instructions
- Rebuild docker image to install new dependency.
- Run an example that has an AOI, while using a `GeoDataWindowConfig` with `method=GeoDataWindowMethod.random`.

Closes #1219

jamesmcclain commented:
:tada:

AdeelH commented on rastervision_pytorch_learner/rastervision/pytorch_learner/learner_config.py:
```diff
@@ -467,6 +467,15 @@ class GeoDataWindowConfig(Config):
         description='Max attempts when trying to find a window within the AOI '
         'of a scene. Only used if method = random and the scene has '
         'aoi_polygons specified.')
+    efficient_aoi_samplings: bool = Field(
```
This is mostly here to provide a way to turn this functionality off on the off chance we encounter an input that the triangulation algorithm cannot handle.

AdeelH replied:
Oops. The 's' at the end is a typo. This supposed to be called `efficient_aoi_sampling` not `efficient_aoi_samplings`.

lewfish commented on tests/pytorch_learner/dataset/test_aoi_sampler.py:
```diff
@@ -0,0 +1,69 @@
+        counts = np.array(
+            [len(block.intersection(points)) for block in blocks])
+
+        p_value = chisquare(counts).pvalue
```
Nice use of a hypothesis test.

lewfish reviewed (approved):
Nice addition! This is approved assuming you manually ran the style tests, unit tests, and integration tests locally, since CI is currently not running (but see https://github.com/azavea/raster-vision/issues/1220) Note that the semseg integration tests are failing https://github.com/azavea/raster-vision/issues/1220

AdeelH commented:
@lewfish One important change since the review:

I've reverted the commit that applied the inverse affine transform in the GeoDataset. The reason that was needed in the first place was that `RasterioCRSTransformer.from_dataset` used to return an `IdentityCRSTransformer` when the `.tif` did not have a CRS even if it did have an affine transform (which was the case with the other project's data). Once I modified `RasterioCRSTransformer.from_dataset` to handle the case of no-CRS + non-identity-transform, everything else just started working as expected.

I'm okay with waiting till the CI is fixed before merging this.

lewfish commented:
Can you do something to trigger CI on this PR? It looks like you force pushed yesterday but I'm not sure if that was after the CI PR was merged.
